feat(hogql): type system

Gilbert09 opened the pull request:
## Problem
- Continuation of the hogql type system from https://github.com/PostHog/posthog/pull/21452

## Changes
- Got all the hogql tests passing ✅ 
- Allowed `ast.Call` functions to fallback to `Unknown` types for args/return types (as it currently is in prod). Allowing us to roll this out with partial completeness of function signatures

#### Notes
- I am concerned about how complete/incomplete these function signatures will be. There will almost certainly be missed signatures which _may_ break something somewhere for someone. How do we feel about this?

## Does this work well for both Cloud and self-hosted?
Yep

## How did you test this code?
Unit tests

EDsCODE committed:
restore wip

EDsCODE committed:
Merge branch 'master' into hogql-func-types

EDsCODE committed:
add numeric operator signatures

EDsCODE committed:
Merge branch 'hogql-func-types' of github.com:PostHog/posthog into hogql-func-types

Gilbert09 committed:
Merge branch 'master' into hogql-func-types

Gilbert09 committed:
Fix all the unit tests

Gilbert09 commented:
@mariusandra I took what we already had and _made it work_. Let me know if there's anything else that **needs** adding before releasing

mariusandra commented on mypy-baseline.txt:
```diff
@@ -532,6 +545,38 @@ posthog/hogql/test/test_resolver.py:0: error: "FieldOrTable" has no attribute "f
 posthog/hogql/test/test_resolver.py:0: error: Item "None" of "JoinExpr | None" has no attribute "table"  [union-attr]
 posthog/hogql/test/test_resolver.py:0: error: Argument 1 to "clone_expr" has incompatible type "SelectQuery | SelectUnionQuery | Field | Any | None"; expected "Expr"  [arg-type]
 posthog/hogql/test/test_resolver.py:0: error: Item "None" of "JoinExpr | None" has no attribute "alias"  [union-attr]
+posthog/hogql/test/test_resolver.py:0: error: Item "None" of "JoinExpr | None" has no attribute "table"  [union-attr]
+posthog/hogql/test/test_resolver.py:0: error: Item "SelectUnionQuery" of "SelectQuery | SelectUnionQuery | Field | Any | None" has no attribute "select"  [union-attr]
+posthog/hogql/test/test_resolver.py:0: error: Item "Field" of "SelectQuery | SelectUnionQuery | Field | Any | None" has no attribute "select"  [union-attr]
+posthog/hogql/test/test_resolver.py:0: error: Item "None" of "SelectQuery | SelectUnionQuery | Field | Any | None" has no attribute "select"  [union-attr]
+posthog/hogql/test/test_resolver.py:0: error: Item "None" of "Type | Any | None" has no attribute "resolve_constant_type"  [union-attr]
+posthog/hogql/test/test_resolver.py:0: error: Item "None" of "Type | None" has no attribute "resolve_constant_type"  [union-attr]
+posthog/hogql/test/test_resolver.py:0: error: Item "None" of "JoinExpr | None" has no attribute "table"  [union-attr]
+posthog/hogql/test/test_resolver.py:0: error: Item "SelectUnionQuery" of "SelectQuery | SelectUnionQuery | Field | Any | None" has no attribute "select"  [union-attr]
+posthog/hogql/test/test_resolver.py:0: error: Item "Field" of "SelectQuery | SelectUnionQuery | Field | Any | None" has no attribute "select"  [union-attr]
+posthog/hogql/test/test_resolver.py:0: error: Item "None" of "SelectQuery | SelectUnionQuery | Field | Any | None" has no attribute "select"  [union-attr]
+posthog/hogql/test/test_resolver.py:0: error: Item "None" of "Type | Any | None" has no attribute "resolve_constant_type"  [union-attr]
+posthog/hogql/test/test_resolver.py:0: error: Item "None" of "Type | None" has no attribute "resolve_constant_type"  [union-attr]
+posthog/hogql/test/test_resolver.py:0: error: Item "None" of "Type | None" has no attribute "resolve_constant_type"  [union-attr]
+posthog/hogql/test/test_resolver.py:0: error: Name "node" already defined on line 0  [no-redef]
+posthog/hogql/test/test_resolver.py:0: error: Item "None" of "Type | None" has no attribute "resolve_constant_type"  [union-attr]
+posthog/hogql/test/test_resolver.py:0: error: Name "node" already defined on line 0  [no-redef]
+posthog/hogql/test/test_resolver.py:0: error: Item "None" of "Type | None" has no attribute "resolve_constant_type"  [union-attr]
+posthog/hogql/test/test_resolver.py:0: error: Name "node" already defined on line 0  [no-redef]
+posthog/hogql/test/test_resolver.py:0: error: Item "None" of "Type | None" has no attribute "resolve_constant_type"  [union-attr]
+posthog/hogql/test/test_resolver.py:0: error: Name "node" already defined on line 0  [no-redef]
+posthog/hogql/test/test_resolver.py:0: error: Item "None" of "Type | None" has no attribute "resolve_constant_type"  [union-attr]
+posthog/hogql/test/test_resolver.py:0: error: Item "None" of "Type | None" has no attribute "resolve_constant_type"  [union-attr]
+posthog/hogql/test/test_resolver.py:0: error: Name "node" already defined on line 0  [no-redef]
+posthog/hogql/test/test_resolver.py:0: error: Item "None" of "Type | None" has no attribute "resolve_constant_type"  [union-attr]
+posthog/hogql/test/test_resolver.py:0: error: Item "None" of "Type | None" has no attribute "resolve_constant_type"  [union-attr]
+posthog/hogql/test/test_resolver.py:0: error: Name "node" already defined on line 0  [no-redef]
+posthog/hogql/test/test_resolver.py:0: error: Item "None" of "Type | None" has no attribute "resolve_constant_type"  [union-attr]
+posthog/hogql/test/test_resolver.py:0: error: Name "node" already defined on line 0  [no-redef]
+posthog/hogql/test/test_resolver.py:0: error: Item "None" of "Type | None" has no attribute "resolve_constant_type"  [union-attr]
+posthog/hogql/test/test_resolver.py:0: error: Item "None" of "Type | None" has no attribute "resolve_constant_type"  [union-attr]
+posthog/hogql/test/test_resolver.py:0: error: Name "node" already defined on line 0  [no-redef]
+posthog/hogql/test/test_resolver.py:0: error: Item "None" of "Type | None" has no attribute "resolve_constant_type"  [union-attr]
```
uh... we do try to reduce lines in this file 😅

mariusandra commented on posthog/hogql/ast.py:
```diff
@@ -303,15 +312,23 @@ def has_child(self, name: str, context: HogQLContext) -> bool:
+    data_type: ConstantDataType = field(default="numeric", init=False)
+
+    def print_type(self) -> str:
+        return "Numeric"
```
Should we do a `raise "should always be float or int"` here to prevent awkward bugs (and introduce few new ones)?

Context: I hate JavaScript's "numeric" type with a passion, and I don't want to introduce one to HogQL if we can at all avoid it. It just makes everything more complex. Keeping it to just "float" and "int" is more than enough. If this is an abstract class that's only used for signatures, then I have nothing against it (let's just make it unusable directly)... but even then, it might really matter for some functions if it's a Int or a Float, and perhaps we should just be really explicit with them in the signatures, e.g. when computing the return type?

mariusandra commented on posthog/hogql/ast.py:
```diff
@@ -361,7 +378,7 @@ def print_type(self) -> str:
 class ArrayType(ConstantType):
     data_type: ConstantDataType = field(default="array", init=False)
-    item_type: ConstantType
+    item_type: Optional[ConstantType] = None  # none means any type
```
do we separate between `any` and `UnknownType`? Should it just be the latter instead of None?

Gilbert09 replied:
We dont - will change to Unknown as default

mariusandra commented on posthog/hogql/ast.py:
```diff
@@ -424,20 +445,47 @@ def is_nullable(self, context: HogQLContext) -> bool:
+        while True:
+            if isinstance(table_type, SelectQueryAliasType):
+                table_type = table_type.select_query_type
+                continue
+
+            if isinstance(table_type, SelectQueryType):
+                field_type = table_type.columns.get(self.name)
+                if field_type is None:
+                    raise QueryError(f"Cant get column: {self.name}")
+
+                constant_type = field_type.resolve_constant_type(context)
+                if constant_type is None:
+                    raise QueryError(f"Constant type cant be resolved: {field_type}")
+
+                return constant_type
+
+            if isinstance(table_type, BaseTableType):
+                table: Table = table_type.resolve_database_table(context)
+                database_field = None
+                if table is not None:
+                    database_field = table.get_field(self.name)
+                if isinstance(database_field, DatabaseField):
+                    constant_type = database_field.get_constant_type()
+                    return constant_type
+                else:
+                    raise NotImplementedError(
+                        f"FieldType.resolve_constant_type, for BaseTableType: unknown database_field type: {str(database_field.__class__)}"
+                    )
+
+            if isinstance(table_type, SelectUnionQueryType):
+                table_type = table_type.types[0]
+                continue
+
+            if isinstance(table_type, SelectViewType):
+                table_type = table_type.select_query_type
+                continue
+
+            raise NotImplementedError(
+                f"FieldType.resolve_constant_type found unexpected table_type: {str(table_type.__class__)}"
+            )
```
It feels like we shouldn't be doing this here... it should be more that we send it off to `table_or_select_or_alias_type.resolve_column_constant_type(self.name)`, and using class inheritance or whatnot, it'll recursively find its way to the deepest node.

mariusandra commented on posthog/hogql/constants.py:
```diff
@@ -5,6 +5,7 @@
 from pydantic import ConfigDict, BaseModel
 
 ConstantDataType: TypeAlias = Literal[
+    "numeric",
```
😬

mariusandra reviewed (commented):
Very cool! Thanks for taking this on.

If all tests pass, I'm happy to get it in and see what's what. I don't think we should worry too much about invalid signatures at this point... as long as whatever we add seems solid, it should be fine.

However, I did leave one bigger comment about the `Numeric` type. If at all possible, I'd completely eradicate it, and just leave `Float` and `Integer`. The numeric concept just adds noise. For example, `plus(1,2)` should return an `Integer` not `Numeric`, because otherwise we can't use it in some places that strictly require integers.

Gilbert09 committed:
Improvements to type system

Gilbert09 commented:
@mariusandra ready for round 2 of review:
- Changed the function signatures to use strict ordering for arguments (it was `any` order before - e.g. it didn't matter what the order of the args were, as long as there was the correct number of matching arg types)
  - Added tests for this
- Fixed a bunch of mypy errors
- Changed `resolve_constant_type` to always return a type
- Stopped mismatched signatures causing queries to error out
- Removed any usage of `Numeric`
- Updated some of the arithmetic signatures
- Implemented `resolve_column_constant_type` for things like `SelectQueryType`
- Pushed the resolving logic down into the deepest node

Gilbert09 committed:
Updatd mypy

Gilbert09 committed:
Merge branch 'master' into hogql-func-types

Gilbert09 committed:
Fixed mypy issues

Gilbert09 committed:
Fixed property types

Gilbert09 committed:
Fixed mypy issues

mariusandra reviewed (approved):
Sweet! This definitely makes life better. Now we just need a lot of definitions and.... 😅 

I have some thoughts inline, but none really blocking, so going to give the ✅. I also do not suggest merging this on a Friday, as who knows what might happen... some runaway recursive loop will take down heavy funnels for example 🤷 😄

EDsCODE commented on posthog/hogql/functions/mapping.py:
```diff
@@ -69,21 +107,189 @@ class HogQLFunctionMeta:
+    ),
     "multiplyDecimal": HogQLFunctionMeta("multiplyDecimal", 2, 3),
     "divideDecimal": HogQLFunctionMeta("divideDecimal", 2, 3),
     # arrays and strings common
```
NIT: add comment here that these signatures are incomplete/what's required to add them—just put in the signature or have to add a test for each one etc

EDsCODE commented on posthog/hogql/test/test_resolver.py:
```diff
@@ -420,3 +420,80 @@ def test_visit_hogqlx_tag_source(self):
         )
 
         assert hogql == expected
```
Tests on actual function call not just arithmetic operations?
